common/variable_ticker.go

```diff
@@ -0,0 +1,315 @@
+
+	// If the current frequency is not equal to the target frequency, the frequency will move towards the
+	// target frequency at this rate per second. If zero, the  ticker will immediately adopt its target frequency.
+	acceleration float64
+
+	// The current frequency of the ticker, in HZ.
+	currentFrequency float64
+
+	// Matches currentFrequency. currentFrequency is the "source of truth", but we cache the period to avoid
+	// recomputing it each tick.
+	currentPeriod time.Duration
+
+	// The previous frequency held by this ticker the last time its configuration was changed.
+	anchorFrequency float64
+
+	// The time at which the ticker last had its configuration changed.
+	anchorTime time.Time
+
+	// The channel that produces an output every time the ticker ticks.
+	tickChan chan struct{}
+
+	// The channel used to send control messages to main ticker loop.
+	controlChan chan any
+}
+
+// frequencyUpdate is a control message to update the target frequency of the ticker.
+type frequencyUpdate struct {
+	// The target frequency to move towards.
+	targetFrequency float64
+}
+
+// accelerationUpdate is a control message to update the acceleration of the ticker.
+type accelerationUpdate struct {
+	// The acceleration to apply to the ticker.
+	acceleration float64
+}
+
+// NewVariableTickerWithPeriod creates a new VariableTicker given a target period.
+func NewVariableTickerWithPeriod(ctx context.Context, period time.Duration) (*VariableTicker, error) {
+	if period <= 0 {
+		return nil, fmt.Errorf("period must be positive, got %v", period)
+	}
+	frequency := float64(time.Second) / float64(period)
+	return NewVariableTickerWithFrequency(ctx, frequency)
+}
+
+// NewVariableTickerWithFrequency creates a new VariableTicker given a target frequency.
+func NewVariableTickerWithFrequency(ctx context.Context, frequency float64) (*VariableTicker, error) {
+	if frequency < 0 {
+		return nil, fmt.Errorf("frequency must be non-negative, got %v", frequency)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	currentPeriod := time.Duration(0)
+	if frequency > 0 {
+		currentPeriod = time.Duration(float64(time.Second) / frequency)
+	}
+
+	ticker := &VariableTicker{
+		ctx:              ctx,
+		close:            cancel,
+		acceleration:     0.0,
+		currentFrequency: frequency,
+		currentPeriod:    currentPeriod,
+		targetFrequency:  frequency,
+		tickChan:         make(chan struct{}),
+		controlChan:      make(chan any, 2),
+	}
+
+	go ticker.run()
+
+	return ticker, nil
+}
+
+// SetTargetPeriod sets the target period for the ticker. If acceleration is non-zero, the ticker will
+// move towards the target period at the rate of acceleration per second. If acceleration is zero,
+// the ticker will immediately adopt the target period.
+func (t *VariableTicker) SetTargetPeriod(period time.Duration) error {
+	if period <= 0 {
+		return fmt.Errorf("invalid period %v, period must be positive", period)
+	}
+
+	frequency := float64(time.Second) / float64(period)
```

Comment:
Any reason not to just pass this `frequency` value into `SetTargetFrequency`, instead of duplicating the min and max checks?

Reply:
good point, simplified